src/scan/postgres_table_reader.cpp

```diff
@@ -0,0 +1,177 @@
+	raw_parsetree_list = pg_parse_query(table_scan_query);
+	List *query_list =
+	    pg_analyze_and_rewrite_fixedparams(linitial_node(RawStmt, raw_parsetree_list), table_scan_query, NULL, 0, NULL);
+	PlannedStmt *planned_stmt = standard_planner(linitial_node(Query, query_list), table_scan_query, 0, NULL);
+	table_scan_query_desc = CreateQueryDesc(planned_stmt, table_scan_query, GetActiveSnapshot(), InvalidSnapshot,
+	                                        None_Receiver, NULL, NULL, 0);
+	ExecutorStart(table_scan_query_desc, 0);
+
+	table_scan_planstate = ExecInitNode(planned_stmt->planTree, table_scan_query_desc->estate, 0);
+
+	table_scan_query_desc->planstate->plan->parallel_aware = true;
+
+	int parallel_workers = ParalleWorkerNumber(cardinality);
+
+	RESUME_CANCEL_INTERRUPTS();
+
+	ParallelContext *pcxt;
+	pei = ExecInitParallelPlan(table_scan_planstate, table_scan_query_desc->estate, NULL, parallel_workers, -1);
+	pcxt = pei->pcxt;
+	LaunchParallelWorkers(pcxt);
+	/* We save # workers launched for the benefit of EXPLAIN */
+	nworkers_launched = pcxt->nworkers_launched;
+
+	/* Set up tuple queue readers to read the results. */
+	if (pcxt->nworkers_launched > 0) {
+		ExecParallelCreateReaders(pei);
+		/* Make a working array showing the active readers */
+		nreaders = pcxt->nworkers_launched;
+		parallel_worker_readers = (void **)palloc(nreaders * sizeof(TupleQueueReader *));
+		memcpy(parallel_worker_readers, pei->reader, nreaders * sizeof(TupleQueueReader *));
+	}
+
+	HOLD_CANCEL_INTERRUPTS();
+
+	slot = ExecInitExtraTupleSlot(table_scan_query_desc->estate, table_scan_planstate->ps_ResultTupleDesc, &TTSOpsMinimalTuple);
+
+	restore_stack_base(current_stack);
+}
+
+PostgresTableReader::~PostgresTableReader() {
+	std::lock_guard<std::mutex> lock(DuckdbProcessLock::GetLock());
+
+	ExecEndNode(table_scan_planstate);
+
+	if (pei != NULL)
+		ExecParallelFinish(pei);
+
+	if (parallel_worker_readers) {
+		pfree(parallel_worker_readers);
+	}
+
+	parallel_worker_readers = NULL;
+
+	ExecParallelCleanup(pei);
+
+	pei = NULL;
+
+	ExecutorFinish(table_scan_query_desc);
+	ExecutorEnd(table_scan_query_desc);
+	FreeQueryDesc(table_scan_query_desc);
+}
+
+/*
+ * This is simple calculation how much postgresql workers we will wanted to spawn for this
+ * postgres table scan.
+ */
+int
+PostgresTableReader::ParalleWorkerNumber(Cardinality cardinality) {
+	static const int base_log = 8;
+	int cardinality_log = std::log2(cardinality);
+	int base = cardinality_log / base_log;
+	return std::max(1, std::min(base, std::min(max_workers_per_postgres_scan, max_parallel_workers)));
+}
+
+TupleTableSlot *
+PostgresTableReader::GetNextTuple() {
+	MinimalTuple worker_minmal_tuple;
+	if (nreaders > 0) {
+		worker_minmal_tuple = GetNextWorkerTuple();
+		if (HeapTupleIsValid(worker_minmal_tuple)) {
+			ExecStoreMinimalTuple(worker_minmal_tuple, slot, false);
+			return slot;
+		}
+	} else {
+		std::lock_guard<std::mutex> lock(DuckdbProcessLock::GetLock());
```

Comment:
Instead of taking the lock for every tuple I expect it might actually result in better performance to take the lock to fill a whole `DataChunk`, so in `PostgresScanTableFunction::PostgresScanFunction`.

That way we're not releasing and acquiring the lock so often. Obviously it has the downside that the lock is held for a longer time. So if you're reading from multiple postgres tables, one of those threads would be paused for a longer time. But I expect that's actually better for performance. Because with these very short locks we're switching back and forth between those threads, and that probably results in lots of thrashing of the L1 cache etc.